### ISSUE TITLE
stdout contention between Log() and Print plugin

### ISSUE THREAD

**Issue by jwestfall69** [pmacct/pmacct#15]
Hi

I originally observed this with 1.5.2 but just confirmed its also happens with master from a few minutes ago.  Log( ) and the Print plugin can both be writing to stdout at the same time, which can result in corrupt lines.  Its easiest to trigger the issue with debug enabled since it generates a large number of Log() lines.

Here is an example:

``` perl
$ cat pmacct.pl 
#!/usr/bin/perl
use strict;
use JSON;

my $count = 0;
while(my $line = <STDIN>) {
  $count++;
  next if($line =~ /^(DEBUG|INFO|WARN)/);
  eval { my $derp = from_json($line); };
  print "JSON ERROR[$count]: $line" if($@);
}
```

```
 $ ./nfacctd -P print -O json -r 5 -l 2557 -d | perl pmacct.pl 
JSON ERROR[2420]: {"bytes": 52, "ip_src": "70.193.DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320330]
JSON ERROR[2421]: 211.140", "packets": 1}
JSON ERROR[2494]: {"bytes": 200, "ip_src": "204.124.15.67", "packDEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320331]
JSON ERROR[2547]: ets": 1}
JSON ERROR[2621]: {"bytes": 56, "ip_src": "74.81.98.247",DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320355]
JSON ERROR[2627]:  "packets": 1}
JSON ERROR[2701]: {"bytes": 253, "ip_src":DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320361]
JSON ERROR[2705]:  "209.254.40.254", "packets": 1}
JSON ERROR[2778]: {"bytes": 40, "ip_src": "76.65.204.120", "paDEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320365]
JSON ERROR[2782]: ckets": 1}
JSON ERROR[9020]: {"bytes": 71, "DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.143:21978] version [9] seqno [4579318]
JSON ERROR[9056]: ip_src": "70.210.48.180", "packets": 1}
JSON ERROR[13405]: {"bytes": 60, "DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320556]
JSON ERROR[13412]: ip_src": "172.56.26.245", "packets": 1}
JSON ERROR[13485]: {"bytes": 126, "ip_src": "181.211.161.124", "packets": DEBUG ( default/core ): Received NetFlow/IPFIX packet from [64.94.0.158:6319] version [9] seqno [157320563]
JSON ERROR[13521]: 2}
```

**Comment by jwestfall69** [comment]
This multiple people writing to stdout also effects just the print plugin itself.  Since multiple print plugin writers can be spawned if thing are getting backed up.  The multiple writers writing to stdout at the same time can result in corrupt lines.

**Comment by paololucente** [comment]
Dear Jim,

Thanks for raising this. There are a couple of possible way out: 1) just document the potential issue in documentation, like: if you point multiple print plugins to stdout and/or log then you will likely run into corruption; 2) work out a locking scheme to prevent this happening (should not be a big deal). My doubt is that while this is OK to prevent corruption of the output of any two print plugins, it would be unacceptable to let a Log() call to wait large outputs by any print plugin to finish.

I'd be inclined towards solution #1 as in my mind pointing things (log, print plugins output) to stdout would be in principle non production approach. Would you have some specific use-case to share that goes against this? I'd be happy to hear.

Cheers,
Paolo

**Comment by jwestfall69** [comment]
Hi

Our end goal is to inject flow data as json into amqp.  Originally we were using the amqp plugin but doing each flow in its own message was super inefficient.  We have some instances that are doing 5000+ flows a second and this ended up causing high cpu usage for everyone involved.

We ended up switching to having an app consume the output of print json, then send the flows to amqp in batches every 250ms.

Another option could be to have Log() print to stderr, then add locking to the print plugins. 

Also it might not have been clear in my previous post.  You can have Log() goto a logfile and a single print plugin enabled and still get the corruption.   If the consumer of the print output is being slow, flows will start getting backed up within nfacctd/pmacct.  From testing this results in additional print writers forking.  When things become unblocked all the forked writers write at once.

You can simulate this issue with a modified script from my original post

``` perl#!/usr/bin/perl
use strict;
use JSON;

my $count = 0;
while(my $line = <STDIN>) {
  $count++;
  eval { my $derp = from_json($line); };
  print "JSON ERROR[$count]: $line" if($@);
  if(-f "sleep.txt") {
    sleep(1);
  }
}
```
start out with sleep.txt existing so things get backed up causing multiple writers to fork.  Then remove sleep.txt and there will be a bunch of corruption.

**Comment by paololucente** [comment]
Hi Jim,

Wrt your original issue please also consider using amqp_multi_values, precisely to avoid creating one flow (or aggregate depending on your aggregation scheme) per message.

The other case where corruption arises is clear. Let me think over a bit; move log to stderr and lock stdout to serialize print plugins output sounds a good option.

Cheers,
Paolo

**Comment by jwestfall69** [comment]
I will look into the amqp_multi_values option, thanks

One of the side benefits of the print json | app method is being able to tweak the json before it gets pushed into amqp.

**Comment by paololucente** [comment]
Hi Jim,

As you can see there have been a few commits referencing this issue. With those, while remaining backward compatible (ie. the title of a CSV is presented only once at the beginning), the two moves you correctly suggested are now implemented: log goes to stderr, stdout is locked by the print plugin. Things seem to test good to my eyes; let me know if things look good to you as well.

Cheers,
Paolo

**Comment by jwestfall69** [comment]
Hi

Both seem to be failing for me.

For log goes to stderr it looks like you changed your fflush to stderr, but didn't actually change your print.
vprintf(msg, ap);
would need to be switched to
vfprintf(stderr, msg, ap);

For the locking I still get a bunch of corruption when doing the sleep test.  I added a debug statement after the file_lock/file_unlock to see when a lock was acquired and released.

```
# ./nfacctd -P print -O json -r 5 -l 2550  | perl pmacct.pl
INFO ( default/core ): Reading configuration from cmdline.
WARN ( default/print ): defaulting to SRC HOST aggregation.
INFO ( default/core ): waiting for NetFlow data on :::2550
INFO ( default/print ): cache entries=16411 base cache memory=53434216 bytes
INFO ( default/print ): *** Purging cache - START (PID: 20233) ***
file_lock: 0, pid: 20233
file_unlock: 0, pid: 20233
INFO ( default/print ): *** Purging cache - END (PID: 20233, QN: 5276/5276, ET: 0) ***
INFO ( default/print ): *** Purging cache - START (PID: 20241) ***
file_lock: 0, pid: 20241
file_unlock: 0, pid: 20241
INFO ( default/print ): *** Purging cache - END (PID: 20241, QN: 8409/8409, ET: 0) ***
INFO ( default/print ): *** Purging cache - START (PID: 20461) ***
file_lock: 0, pid: 20461
file_unlock: 0, pid: 20461
INFO ( default/print ): *** Purging cache - END (PID: 20461, QN: 8842/8842, ET: 0) ***
```
Things look sane with a single print worker, however if I touch sleep.txt to cause pmacct.pl to stall, the results show no locking actually happening.

```
INFO ( default/print ): *** Purging cache - START (PID: 20470) ***
file_lock: 0, pid: 20470
INFO ( default/print ): *** Purging cache - START (PID: 20481) ***
file_lock: 0, pid: 20481
INFO ( default/print ): *** Purging cache - START (PID: 26740) ***
file_lock: 0, pid: 26740
INFO ( default/print ): *** Purging cache - START (PID: 26746) ***
file_lock: 0, pid: 26746
INFO ( default/print ): *** Purging cache - START (PID: 26755) ***
file_lock: 0, pid: 26755
INFO ( default/print ): *** Purging cache - START (PID: 27060) ***
file_lock: 0, pid: 27060
INFO ( default/print ): *** Purging cache - START (PID: 27082) ***
file_lock: 0, pid: 27082
INFO ( default/print ): *** Purging cache - START (PID: 27091) ***
file_lock: 0, pid: 27091
INFO ( default/print ): *** Purging cache - START (PID: 27099) ***
file_lock: 0, pid: 27099
INFO ( default/print ): *** Purging cache - START (PID: 27106) ***
file_lock: 0, pid: 27106
WARN ( default/print ): Maximum number of writer processes reached (10).
WARN ( default/print ): Maximum number of writer processes reached (10).
```

If locking was working correctly only the first worker would have been able to acquire the lock, but the above output shows all of the workers are getting the lock.  I think the issue here is that its not possible to flock on stdout.

As I test I had the workers open and flock a file in /tmp, and it resulted in the correct behavior.

```
INFO ( default/print ): *** Purging cache - START (PID: 5848) ***
file_lock: 0, pid: 5848
INFO ( default/print ): *** Purging cache - START (PID: 5857) ***
INFO ( default/print ): *** Purging cache - START (PID: 5862) ***
INFO ( default/print ): *** Purging cache - START (PID: 6064) ***
INFO ( default/print ): *** Purging cache - START (PID: 6202) ***
INFO ( default/print ): *** Purging cache - START (PID: 6231) ***
INFO ( default/print ): *** Purging cache - START (PID: 6245) ***
INFO ( default/print ): *** Purging cache - START (PID: 6269) ***
INFO ( default/print ): *** Purging cache - START (PID: 6281) ***
INFO ( default/print ): *** Purging cache - START (PID: 6484) ***
WARN ( default/print ): Maximum number of writer processes reached (10).
```

This seemed to correct 99% of the corruption when I rm'd sleep.txt.  I think the remaining 1% will be fixed by doing a fflush(f) before releasing the lock.  

thanks
jim

**Comment by paololucente** [comment]
Hi Jim,

Thanks for your feedback. I will make one more review of all of this. One thing though: the patch in log.c seems quite straightforward - why don't you put a PR for it yourself? It would give you full credit for it.

The other proposal, ie. using a file lock for stdout, also looks sane and straightforward to me but will likely need the introduction of a new configuration directive - i can take care of all of that.

Cheers,
Paolo

**Comment by paololucente** [comment]
Hi Jim,

Your PR was merged and a new print_output_lock_file config directive is now introduced for selecting a lock file on disk. I confirm testing works well for me. Please let me know if it does for you as well.

Cheers,
Paolo

**Comment by jwestfall69** [comment]
Hi

I did another PR to fix one last thing.  That plus the print_output_lock_file bits fixes everything for me.

thanks!
jim

**Comment by paololucente** [comment]
Great stuff. Just a note: i so far don't see any new PR from you.

**Comment by jwestfall69** [comment]
Weird, somehow I did a PR from me to my fork.  You should see it now.

**Comment by paololucente** [comment]
Perfect Jim. Just merged your PR and going to close this issue. Thanks for all the support.

Paolo
